ellipsoid/v1/caesar/spheroid.hpp

```diff
@@ -0,0 +1,178 @@
+#pragma once
+
+#include <cmath>
+
+namespace caesar {
+
+/**
+ * An ellipsoid with two axes of equal length
+ *
+ * A spheroid is parameterized by its semi-major and semi-minor axis lengths,
+ * \f$a\f$ and \f$b\f$ as
+ *
+ * \f[
+ * \frac{x^2 + y^2}{a^2} + \frac{z^2}{b^2} = 1
+ * \f]
+ *
+ * In geodesy, a spheroid is typically used as a geodetic reference ellipsoid
+ * that approximates the size and shape of the Earth. In such applications, the
+ * semi-major axis length describes the equatorial radius and the semi-minor
+ * axis length describes the polar radius.
+ *
+ * Reference ellipsoids are commonly specified by their semi-major axis and
+ * flattening constant, \f$f\f$, defined as
+ *
+ * \f[
+ * f = \frac{a - b}{a}
+ * \f]
+ *
+ * The flattening is a measure of the ellipticity of the spheroid. If the
+ * flattening is between 0 and 1, the spheroid is oblate (\f$a > b\f$). If it's
+ * equal to 0, the spheroid is a perfect sphere (\f$a == b\f$). If it's less
+ * than zero, the spheroid is prolate (\f$a < b\f$).
+ *
+ * \see wgs84_ellipsoid
+ */
+class Spheroid {
+public:
+    /**
+     * Construct a new Spheroid object.
+     *
+     * \param[in] a semi-major axis length
+     * \param[in] f flattening constant
+     */
+    constexpr Spheroid(double a, double f) noexcept : a_(a), f_(f) {}
+
+    /**
+     * Return the semi-major axis length.
+     *
+     * In geodetic applications, this is the equatorial radius.
+     *
+     * \see Spheroid::a
+     */
+    constexpr double
+    semimajor_axis() const noexcept
```

Comment:
```suggestion
    [[nodiscard]] constexpr double
    semimajor_axis() const noexcept
```

What are your thoughts on this? It could apply to basically every getter here

Reply:
Yeah, I'm a fan of `[[nodiscard]]` too. I started using it in my code for a while but sometimes gcc will generate lots of warnings that the attribute is being ignored. I never figured out what that was about so I've been on a hiatus from using it until I could get some insight.

Have you run into this issue at all? I'll try to produce a MWE to share

Reply:
Maybe you are thinking of this issue? https://quuxplusone.github.io/blog/2019/01/25/nodiscard-philosophies/

I like [[nodiscard]]. I think the counterargument is it's yet another should-be-default qualifier that you have to sprinkle around liberally, like constexpr and noexcept. But I'm okay with that I think.

Reply:
Oops I think I figured it out. I had the `[[nodiscard]]` attribute after the `constexpr` specifier. That was giving me a compiler warning like this:
```
test.cpp:7:15: warning: attribute ignored [-Wattributes]
    7 |     constexpr [[nodiscard]] double
      |               ^
test.cpp:7:15: note: an attribute that appertains to a type-specifier is ignored
```

Your way (`[[nodiscard]]` before `constexpr`) compiles without warnings :sweat_smile: